### ISSUE TITLE
How to train model   afresh ?

### ISSUE THREAD

**Issue by rhp228228** [hfslyc/AdvSemiSeg#5]
Hi, when I use your code by your README, I find that the model training is fintuning  on your well-trained model. When I choose not to fintuning , and want to retrain the data VOC2012. Some wrongs with the loss. The losses don't  converge. I want to konw if anything wrong with your code? emmm,  I with you to test your code.  Thanks.

**Comment by chuanruihu** [comment]
Hi, I am interesting about it . Do you solve your problem now?

**Comment by hfslyc** [comment]
Hi all, I'm investigating the problem. Something might go wrong when I was cleaning the code.

**Comment by hfslyc** [comment]
btw, @rhp228228 what do you mean by "not to fintuning"?  the model will not work without the imagenet pretrained model.

**Comment by John1231983** [comment]
@wasidennis : I think he means training from scratch. without using imagenet pretrain model. I am worried about why the model cannot train from scratch. If I remove the Adversarial network, the segmentation network still can train from scratch although the performance is not perfect. However, it at least provides a meaningful accuracy. One thing I suggest for training from scratch is that train the Segmentation network with certain iterations (about 1000 iterations), then train Adversarial network. How do you think?

This is what I train your model from scratch with my own dataset: First, the loss seg, adv, loss D are reduced by iterations. However, after certain iterations (likes 100 iters), the loss adv increases (other loss still decreases). What is happen? I have tried to reduce the learning rate but the problem still occurs.

**Comment by hfslyc** [comment]
@John1231983  Is there any specific reason why you do not use a pre-trained model? most dataset would benefit from an Imgenet pretrained model.

If you insist to train from scratch, it will take a while for the segmentation network to produce meaningful results. In this case, applying adversarial learning too early might causing the gradient explosion since D could easily discriminate the network result and groundtruth, resulting in a huge adversarial loss.

**Comment by John1231983** [comment]
Could I ask you one more question? I have to train from scratch because my own dataset is different with imagenet. I have success to train it althought adv loss increase from 0.7 to 21 after 10.000 epochs. After add sigmod function to adv, my loss increase from 0.3 to 0.8 after 2000 epochs and did not change from 2000 to 10000 epochs. Could you tell me why you did not use sigmod in the end of discriminator network ( or why not use sigmod in adv, while semi loss use sigmod?)? Thanks

**Comment by hfslyc** [comment]
The sigmoid function is integrated into the binary cross entropy loss function.

**Comment by John1231983** [comment]
does it integrated in your binary cross entropy or official binanry cross entropy of pytorch? I have not find the line in your binary cross entropy doing it

**Comment by hfslyc** [comment]
https://github.com/hfslyc/AdvSemiSeg/blob/111d5183a966725d70db0368e9ff4f90ce277756/utils/loss.py#L65

The binary_cross_entropy_with_logits does contain sigmoid.
ref: https://pytorch.org/docs/0.3.0/nn.html?highlight=functional#torch.nn.BCEWithLogitsLoss

**Comment by John1231983** [comment]
Great. One more thing I found that you use the softmax layer, it looks redundant because you already use sigmoid layer in BCEwithlogits
>The output of D in[0,1], which is easy to do if we tack on a sigmoid layer at the end.

What is purpose for using softmax?

**Comment by hfslyc** [comment]
@John1231983 which softmax are you referring to?

**Comment by John1231983** [comment]
@hfslyc  this one in the train.py
```
D_out = interp(model_D(F.softmax(pred)))
```

**Comment by wasidennis** [comment]
Sigmoid is implemented in the loss function but not in the model, i.e., 'pred' is not after softmax. As we also need the output after softmax as the input for the discriminator, we need to use the softmax function on 'pred'.

**Comment by John1231983** [comment]
Thanks wasidennis for it, I understand now, the softmax is to normalize the output of segmentation to [0,1] and used it as input for discriminator network. The sigmoid is for last layer of discriminator

**Comment by John1231983** [comment]
The last question related to the training D. In DCGAN, we will optimize D as
```
############################
# (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
###########################
# train with real
errD_real = criterion(output, label)
errD_real.backward()
# train with fake
errD_fake = criterion(output, label)
errD_fake.backward()
errD = errD_real + errD_fake
optimizerD.step()
```
While, your code for optimize D is [L380](https://github.com/hfslyc/AdvSemiSeg/blob/master/train.py#L380)
```
# train with pred
loss_D = bce_loss(D_out, make_D_label(pred_label, ignore_mask))
loss_D.backward()
# train with gt
loss_D = bce_loss(D_out, make_D_label(gt_label, ignore_mask_gt))
loss_D.backward()
optimizerD.step()
```

My question is why not use sum operator as `errD = errD_real + errD_fake`? With/without the line does not affect to result, right?

**Comment by hfslyc** [comment]
Yes, it will not affect the results.
